File: data/src/main/kotlin/com/almadevelop/comixreader/data/entity/ComicRackMetadata.kt
package com.almadevelop.comixreader.data.entity

import androidx.annotation.Keep
import androidx.room.*

/**
 * @see <a href="https://github.com/ciromattia/kcc/wiki/ComicRack-metadata">ComicRankMetadata</a>
 * @see <a href="https://github.com/dickloraine/EmbedComicMetadata/blob/master/comicinfoxml.py">Parser</a>
 */
@Entity(
    tableName = ComicRackMetadata.TABLE_NAME,
    foreignKeys = [ForeignKey(
        entity = ComicBook::class,
        childColumns = [ComicRackMetadata.COLUMN_BOOK_ID],
        parentColumns = [ComicBook.COLUMN_ID],
        onDelete = ForeignKey.CASCADE
    )],
    indices = [Index(
        name = ComicRackMetadata.INDEX_BOOK_ID,
        value = [ComicRackMetadata.COLUMN_BOOK_ID],
        unique = true
    )]
)
data class ComicRackMetadata @Ignore constructor(
    @PrimaryKey(autoGenerate = true)
    @ColumnInfo(name = COLUMN_ID)
    val id: Long = 0,
    @ColumnInfo(name = COLUMN_BOOK_ID)
    val bookId: Long = 0,
    @ColumnInfo(name = COLUMN_TITLE, defaultValue = "NULL")
    val title: String? = null,
    @ColumnInfo(name = COLUMN_SERIES, defaultValue = "NULL")
    val series: String? = null,
    @ColumnInfo(name = COLUMN_SUMMARY, defaultValue = "NULL")
    val summary: String? = null,
    @ColumnInfo(name = COLUMN_NUMBER, defaultValue = "NULL")
    val number: Int? = null, //issue
    @ColumnInfo(name = COLUMN_COUNT, defaultValue = "NULL")
    val count: Int? = null, //issue count
    @ColumnInfo(name = COLUMN_VOLUME, defaultValue = "NULL")
    val volume: Int? = null, // Title > Volume > Issue
    @ColumnInfo(name = COLUMN_PAGE_COUNT, defaultValue = "NULL")
    val pageCount: Int? = null,
    @ColumnInfo(name = COLUMN_YEAR, defaultValue = "NULL")
    val year: Int? = null,
    @ColumnInfo(name = COLUMN_MONTH, defaultValue = "NULL")
    val month: Int? = null,
    @ColumnInfo(name = COLUMN_DAY, defaultValue = "NULL")
    val day: Int? = null,
    @ColumnInfo(name = COLUMN_PUBLISHER, defaultValue = "NULL")
    val publisher: String? = null,
    @ColumnInfo(name = COLUMN_WRITER, defaultValue = "NULL")
    val writer: String? = null,
    @ColumnInfo(name = COLUMN_PENCILLER, defaultValue = "NULL")
    val penciller: String? = null,
    @ColumnInfo(name = COLUMN_INKER, defaultValue = "NULL")
    val inker: String? = null,
    @ColumnInfo(name = COLUMN_COLORIST, defaultValue = "NULL")
    val colorist: String? = null,
    @ColumnInfo(name = COLUMN_LETTERER, defaultValue = "NULL")
    val letterer: String? = null,
    @ColumnInfo(name = COLUMN_COVER_ARTIST, defaultValue = "NULL")
    val coverArtist: String? = null,
    @ColumnInfo(name = COLUMN_EDITOR, defaultValue = "NULL")
    val editor: String? = null,
    @ColumnInfo(name = COLUMN_IMPRINT, defaultValue = "NULL")
    val imprint: String? = null,
    @ColumnInfo(name = COLUMN_GENRE, defaultValue = "NULL")
    val genre: String? = null,
    @ColumnInfo(name = COLUMN_FORMAT, defaultValue = "NULL")
    val format: String? = null,
    @ColumnInfo(name = COLUMN_AGE_RATING, defaultValue = "NULL")
    val ageRating: String? = null,
    @ColumnInfo(name = COLUMN_TEAMS, defaultValue = "NULL")
    val teams: String? = null,
    @ColumnInfo(name = COLUMN_LOCATIONS, defaultValue = "NULL")
    val locations: String? = null,
    @ColumnInfo(name = COLUMN_STORY_ARC, defaultValue = "NULL")
    val storyArc: String? = null,
    @ColumnInfo(name = COLUMN_SERIES_GROUP, defaultValue = "NULL")
    val seriesGroup: String? = null,
    @ColumnInfo(name = COLUMN_BLACK_WHITE, defaultValue = "NULL")
    val blackAndWhite: Boolean? = null,
    @ColumnInfo(name = COLUMN_MANGA, defaultValue = "NULL")
    val manga: Boolean? = null,
    @ColumnInfo(name = COLUMN_CHARACTERS, defaultValue = "NULL")
    val characters: String? = null,
    @ColumnInfo(name = COLUMN_WEB, defaultValue = "NULL")
    val web: String? = null,
    @ColumnInfo(name = COLUMN_NOTES, defaultValue = "NULL")
    val notes: String? = null,
    @ColumnInfo(name = COLUMN_LANGUAGE, defaultValue = "NULL")
    val languageIso: String? = null,
    @Ignore
    val pages: List<ComicRackPageMetadata>? = null
) {
    /**
     * Used by Room
     */
    @Suppress("unused")
    @Keep
    internal constructor(
        id: Long,
        bookId: Long,
        title: String?,
        series: String?,
        summary: String?,
        number: Int?,
        count: Int?,
        volume: Int?,
        pageCount: Int?,
        year: Int?,
        month: Int?,
        day: Int?,
        publisher: String?,
        writer: String?,
        penciller: String?,
        inker: String?,
        colorist: String?,
        letterer: String?,
        coverArtist: String?,
        editor: String?,
        imprint: String?,
        genre: String?,
        format: String?,
        ageRating: String?,
        teams: String?,
        locations: String?,
        storyArc: String?,
        seriesGroup: String?,
        blackAndWhite: Boolean?,
        manga: Boolean?,
        characters: String?,
        web: String?,
        notes: String?,
        languageIso: String?
    ) : this(
        id,
        bookId,
        title,
        series,
        summary,
        number,
        count,
        volume,
        pageCount,
        year,
        month,
        day,
        publisher,
        writer,
        penciller,
        inker,
        colorist,
        letterer,
        coverArtist,
        editor,
        imprint,
        genre,
        format,
        ageRating,
        teams,
        locations,
        storyArc,
        seriesGroup,
        blackAndWhite,
        manga,
        characters,
        web,
        notes,
        languageIso,
        null
    )

    /**
     * Called from a native side
     */
    @Suppress("unused")
    @Keep
    @Ignore
    internal constructor(
        title: String?,
        series: String?,
        summary: String?,
        number: Int?,
        count: Int?,
        volume: Int?,
        pageCount: Int?,
        year: Int?,
        month: Int?,
        day: Int?,
        publisher: String?,
        writer: String?,
        penciller: String?,
        inker: String?,
        colorist: String?,
        letterer: String?,
        coverArtist: String?,
        editor: String?,
        imprint: String?,
        genre: String?,
        format: String?,
        ageRating: String?,
        teams: String?,
        locations: String?,
        storyArc: String?,
        seriesGroup: String?,
        blackAndWhite: Boolean?,
        manga: Boolean?,
        characters: String?,
        web: String?,
        notes: String?,
        languageIso: String?,
        pages: Array<ComicRackPageMetadata>?
    ) : this(
        0,
        0,
        title,
        series,
        summary,
        number,
        count,
        volume,
        pageCount,
        year,
        month,
        day,
        publisher,
        writer,
        penciller,
        inker,
        colorist,
        letterer,
        coverArtist,
        editor,
        imprint,
        genre,
        format,
        ageRating,
        teams,
        locations,
        storyArc,
        seriesGroup,
        blackAndWhite,
        manga,
        characters,
        web,
        notes,
        languageIso,
        pages?.asList()
    )

    companion object {
        internal const val TABLE_NAME = "comic_rack_metadata"

        internal const val COLUMN_ID = "id"
        internal const val COLUMN_BOOK_ID = "book_id"
        internal const val COLUMN_TITLE = "title"
        internal const val COLUMN_SERIES = "series"
        internal const val COLUMN_SUMMARY = "summary"
        internal const val COLUMN_NUMBER = "number"
        internal const val COLUMN_COUNT = "count"
        internal const val COLUMN_VOLUME = "volume"
        internal const val COLUMN_PAGE_COUNT = "page_count"
        internal const val COLUMN_YEAR = "year"
        internal const val COLUMN_MONTH = "month"
        internal const val COLUMN_DAY = "day"
        internal const val COLUMN_PUBLISHER = "publisher"
        internal const val COLUMN_WRITER = "writer"
        internal const val COLUMN_PENCILLER = "penciller"
        internal const val COLUMN_INKER = "inker"
        internal const val COLUMN_COLORIST = "colorist"
        internal const val COLUMN_LETTERER = "letterer"
        internal const val COLUMN_COVER_ARTIST = "cover_artist"
        internal const val COLUMN_EDITOR = "editor"
        internal const val COLUMN_IMPRINT = "imprint"
        internal const val COLUMN_GENRE = "genre"
        internal const val COLUMN_FORMAT = "format"
        internal const val COLUMN_AGE_RATING = "age_rating"
        internal const val COLUMN_TEAMS = "teams"
        internal const val COLUMN_LOCATIONS = "locations"
        internal const val COLUMN_STORY_ARC = "story_arc"
        internal const val COLUMN_SERIES_GROUP = "series_group"
        internal const val COLUMN_BLACK_WHITE = "black_white"
        internal const val COLUMN_MANGA = "manga"
        internal const val COLUMN_CHARACTERS = "characters"
        internal const val COLUMN_WEB = "web"
        internal const val COLUMN_NOTES = "notes"
        internal const val COLUMN_LANGUAGE = "language"

        internal const val INDEX_BOOK_ID = "idx_book_id"
    }
}

File: comix_tensors/src/android/jni_app/objects/app/comic_rack_metadata.rs
use jni::errors::Result as JniResult;
use jni::objects::{JClass, JObject};
use jni::JNIEnv;
use once_cell::sync::OnceCell;

use crate::{ComicInfo as ComicInfoData, ComicInfoPage as ComicInfoPageData};

use super::{constants::*, java, string_to_jobject, try_cache_class_descr, CacheClassDescr};

static CLASS_DESCR: OnceCell<CacheClassDescr> = OnceCell::new();

pub type ComicRackMetadata<'a> = JObject<'a>;

fn class_descr(env: &JNIEnv) -> &'static CacheClassDescr {
    return CLASS_DESCR.get_or_init(||
        try_cache_class_descr(env,
                              COMIC_RACK_METADATA_TYPE,
                              format!("(Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;Ljava/lang/Integer;Ljava/lang/Integer;Ljava/lang/Integer;Ljava/lang/Integer;Ljava/lang/Integer;Ljava/lang/Integer;Ljava/lang/Integer;Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;Ljava/lang/Boolean;Ljava/lang/Boolean;Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;[L{};)V", COMIC_RACK_PAGE_TYPE).as_str()));
}

pub fn new<'a>(env: &'a JNIEnv, comic_info: &Option<ComicInfoData>) -> JniResult<ComicRackMetadata<'a>> {
    let CacheClassDescr(class, constructor) = class_descr(env);

    match comic_info {
        Some(comic_info) => env.new_object_unchecked(
            JClass::from(class.as_obj()),
            *constructor,
            &[
                string_to_jobject(env, &comic_info.title)?.into(),
                string_to_jobject(env, &comic_info.series)?.into(),
                string_to_jobject(env, &comic_info.summary)?.into(),
                java::integer::new(env, comic_info.number.map(|n| n as _))?.into(),
                java::integer::new(env, comic_info.count.map(|n| n as _))?.into(),
                java::integer::new(env, comic_info.volume.map(|n| n as _))?.into(),
                java::integer::new(env, comic_info.page_count.map(|n| n as _))?.into(),
                java::integer::new(env, comic_info.year.map(|n| n as _))?.into(),
                java::integer::new(env, comic_info.month.map(|n| n as _))?.into(),
                java::integer::new(env, comic_info.day.map(|n| n as _))?.into(),
                string_to_jobject(env, &comic_info.publisher)?.into(),
                string_to_jobject(env, &comic_info.writer)?.into(),
                string_to_jobject(env, &comic_info.penciller)?.into(),
                string_to_jobject(env, &comic_info.inker)?.into(),
                string_to_jobject(env, &comic_info.colorist)?.into(),
                string_to_jobject(env, &comic_info.letterer)?.into(),
                string_to_jobject(env, &comic_info.cover_artist)?.into(),
                string_to_jobject(env, &comic_info.editor)?.into(),
                string_to_jobject(env, &comic_info.imprint)?.into(),
                string_to_jobject(env, &comic_info.genre)?.into(),
                string_to_jobject(env, &comic_info.format)?.into(),
                string_to_jobject(env, &comic_info.age_rating)?.into(),
                string_to_jobject(env, &comic_info.teams)?.into(),
                string_to_jobject(env, &comic_info.locations)?.into(),
                string_to_jobject(env, &comic_info.story_arc)?.into(),
                string_to_jobject(env, &comic_info.series_group)?.into(),
                java::boolean::new(env, comic_info.black_and_white.map(|b| b as _))?.into(),
                java::boolean::new(env, comic_info.manga.map(|b| b as _))?.into(),
                string_to_jobject(env, &comic_info.characters)?.into(),
                string_to_jobject(env, &comic_info.web)?.into(),
                string_to_jobject(env, &comic_info.notes)?.into(),
                string_to_jobject(env, &comic_info.language_iso)?.into(),
                page::new_array(env, &comic_info.pages)?.into(),
            ],
        ),
        None => Ok(JObject::null()),
    }
}

pub mod page {
    use super::*;

    static CLASS_DESCR: OnceCell<CacheClassDescr> = OnceCell::new();

    pub type ComicRackPageMetadata<'a> = JObject<'a>;
    pub type ComicRackPageMetadataArray<'a> = JObject<'a>;

    fn class_descr(env: &JNIEnv) -> &'static CacheClassDescr {
        return CLASS_DESCR.get_or_init(||
            try_cache_class_descr(env,
                                  COMIC_RACK_PAGE_TYPE,
                                  "(Ljava/lang/Integer;Ljava/lang/String;Ljava/lang/Long;Ljava/lang/Integer;Ljava/lang/Integer;)V"));
    }

    pub fn new<'a>(env: &'a JNIEnv, page: &ComicInfoPageData) -> JniResult<ComicRackPageMetadata<'a>> {
        let CacheClassDescr(class, constructor) = class_descr(env);

        env.new_object_unchecked(
            JClass::from(class.as_obj()),
            *constructor,
            &[
                java::integer::new(env, page.image.map(|n| n as _))?.into(),
                string_to_jobject(env, &page.image_type)?.into(),
                java::long::new(env, page.image_size.map(|n| n as _))?.into(),
                java::integer::new(env, page.image_width.map(|n| n as _))?.into(),
                java::integer::new(env, page.image_height.map(|n| n as _))?.into(),
            ],
        )
    }

    pub fn new_array<'a>(
        env: &'a JNIEnv,
        pages: &Option<Vec<ComicInfoPageData>>,
    ) -> JniResult<ComicRackPageMetadataArray<'a>> {
        let CacheClassDescr(class, _) = class_descr(env);

        Ok(match pages {
            Some(pages) => {
                let pages_jni_array = env.new_object_array(
                    pages.len() as _,
                    JClass::from(class.as_obj()),
                    JObject::null(),
                )?;

                for (i, page) in pages.into_iter().enumerate() {
                    let page = new(env, page)?;
                    env.set_object_array_element(pages_jni_array, i as _, page)?;
                }

                JObject::from(pages_jni_array)
            }
            None => JObject::null(),
        })
    }
}
